cmd/thanos/rule.go

```diff
@@ -642,6 +642,21 @@ func runRule(
 		)
 		storeServer := store.NewLimitedStoreServer(store.NewInstrumentedStoreServer(reg, tsdbStore), reg, conf.storeRateLimits)
 		options = append(options, grpcserver.WithServer(store.RegisterStoreServer(storeServer, logger)))
+
+		ctx, cancel := context.WithCancel(context.Background())
```

Comment:
is there a reason we dont use top level context + cancel for all the concurrent routines? I was wondering yesterday; if some top one fails this one wont be canceled. 

I would move ctx, cancel to top i think.

Reply:
I was wrong about the context. The cancellation is handled by the run.Group invoking the interrupt function passed to g.Add, which means that this is canceling its own context which has correct semantics i think. 
Having said that, run.Group is not very context friendly i think.